aten/src/ATen/native/cpu/ReduceOpsKernel.cpp

```diff
@@ -258,19 +257,15 @@ static void norm_kernel_tensor_iterator_impl(
         });
       });
   } else {
-    if (iter.dtype(0) == kHalf) {
-      return norm_kernel_cpu_impl<at::Half, float>(iter, val);
-    } else if (iter.input_dtype() == kHalf && iter.dtype(0) == kFloat) {
+    if (iter.input_dtype() == kHalf && iter.dtype(0) == kFloat) {
       // type promotion that does cast and reduction in a single kernel
-      return norm_kernel_cpu_impl<at::Half, float, float>(iter, val);
-    } else if(iter.dtype(0) == kBFloat16) {
-      return norm_kernel_cpu_impl<at::BFloat16, float>(iter, val);
+      return norm_kernel_cpu_impl<at::Half, float>(iter, val);
     } else if (iter.input_dtype() == kBFloat16 && iter.dtype(0) == kFloat) {
       // type promotion that does cast and reduction in a single kernel
-      return norm_kernel_cpu_impl<at::BFloat16, float, float>(iter, val);
+      return norm_kernel_cpu_impl<at::BFloat16, float>(iter, val);
     }
 
-    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES(iter.input_dtype(), "norm_cpu", [&] {
+    AT_DISPATCH_FLOATING_AND_COMPLEX_TYPES_AND3(kHalf, kBFloat16, kComplexHalf, iter.input_dtype(), "norm_cpu", [&] {
```

Comment:
Are the kHalf and kBfloat16 types in the AND3 redundant? It seemed to support those types before??? or it suggests we aren't testing the right thing.

Reply:
Aren't these dispatched already on 262 and 265?

Reply:
> Are the kHalf and kBfloat16 types in the AND3 redundant? It seemed to support those types before??? or it suggests we aren't testing the right thing.

Because I modified the logic above, now kHalf and kBfloat16 types may also be here (the case of `iter.input_dtype() == kHalf/kBFloat16 and iter.dtype(0)!= kFloat`)

Reply:
I checked the UT is for this operator, but I got this warning when run the UT, ComplexHalf seems not fully supported yet: 
```
/home/jiayisun/pytorch/torch/testing/_creation.py:233: UserWarning: ComplexHalf support is experimental and many operators don't support it yet. (Triggered internally at /home/jiayisun/pytorch/aten/src/ATen/EmptyTensor.cpp:46.)
  result = torch.empty(shape, device=device, dtype=dtype)
Some dtypes for norm on device type cpu are only partially supported!
The following dtypes only worked on some samples during backward: {torch.complex32}.
```